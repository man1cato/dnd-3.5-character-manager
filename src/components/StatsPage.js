import React from 'react';
import {connect} from 'react-redux';
import update from 'react-addons-update';
import {startEditProfile} from '../actions/profile';

import Header from './Header';
import Footer from './Footer';
import Abilities from './Abilities';
import SkillSet from './SkillSet';


export class StatsPage extends React.Component {
  state = {
    abilities: this.props.abilities
  }

  componentWillUnmount() {
    this.props.startEditProfile(this.props.id, {
      abilities: this.state.abilities
    });
  }

  onInputChange = (e) => {
    const ability = e.target.id;
    const value = Number(e.target.value);
    const tempScore = value === 0 || isNaN(value) ? "" : value;
    const tempMod = tempScore === "" ? "" : Math.floor(tempScore/2 - 5);
    this.setState((prevState) => ({
      abilities: update(prevState.abilities, {
        [ability]: {
          tempScore: {$set: tempScore},
          tempMod: {$set: tempMod}
        }
      })
    }))
  }  

  render () {
    return (
      <div className="layout">
        <Header pageTitle="Character Stats" />
        <div className="layout__body">

          <div className="d-flex justify-content-between">
            <h3>Level {this.props.level}</h3>
            <span>XP: {this.props.xp}</span>
          </div>

          <div className="d-flex justify-content-between">
            <span>Hit Die: {this.props.hd}</span>
            <span>To next level: {this.props.toNextLevel}</span>
          </div>

          <Abilities
            abilities={this.state.abilities}
            onInputChange={this.onInputChange}
          />

          <SkillSet 
            skillSet={this.props.skillSet}
          />

        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  uid: state.auth.uid,
  hd: state.profile.hd,
  level: state.profile.level,
  xp: state.profile.xp,
  toNextLevel: state.profile.toNextLevel,
  abilities: state.profile.abilities,
  skillSet: state.profile.skillSet
})

const mapDispatchToProps = (dispatch, props) => ({ 
  startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
});

export default connect(mapStateToProps, mapDispatchToProps)(StatsPage);
