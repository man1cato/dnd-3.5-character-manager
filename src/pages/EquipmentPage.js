import React, { Fragment, useState, useEffect } from 'react'
import { connect } from 'react-redux'
import update from 'immutability-helper'
import _ from 'lodash'

import ItemModal from '../components/ItemModal'
import { startEditProfile } from '../store/actions/profile'
import { 
	calcTotalMoney, calcItemTotalValue, calcItemTotalWeight, 
	calcEquipmentTotalValue, calcEquipmentTotalWeight 
} from '../utils/utils'


// const calcEquipmentTotalValue = (equipment) => equipment.map((item) => item.totalValue).reduce((total, num) => total + num)
// const calcEquipmentTotalWeight = (equipment) => equipment.map((item) => item.totalWeight).reduce((total, num) => total + num)

const denominations = ['pp', 'gp', 'sp', 'cp']

export const EquipmentPage = (props) => {
	const [money, setMoney] = useState(props.money)
	const [totalMoney, setTotalMoney] = useState(calcTotalMoney(money))
	
	const [equipment, setEquipment] = useState(props.equipment.map((item) => ({
		...item,
		totalValue: calcItemTotalValue(props.items[item.id], item.qty),
		totalWeight: calcItemTotalWeight(props.items[item.id], item.qty)
	})) || [])

	const [equipmentTotalValue, setEquipmentTotalValue] = useState(calcEquipmentTotalValue(equipment, props.items) || 0)
	const [equipmentTotalWeight, setEquipmentTotalWeight] = useState(calcEquipmentTotalWeight(equipment, props.items) || 0)

	const [equipped, setEquipped] = useState({
		weapons: props.equipped.weapons || [],
		armor: props.equipped.armor || null,
		shield: props.equipped.shield || null
	})
	
	const [selected, setSelected] = useState(undefined)

	useEffect(() => {
		setTotalMoney(calcTotalMoney(money))
		props.startEditProfile(props.id, { money })
	}, [money])

	useEffect(() => {
		setEquipmentTotalValue(calcEquipmentTotalValue(equipment, props.items))
		setEquipmentTotalWeight(calcEquipmentTotalWeight(equipment, props.items))
		props.startEditProfile(props.id, { equipment: _.map(equipment, (item) => ({ id: item.id, qty: item.qty })) })
	}, [equipment])

	const handleChange = (e) => {
		const id = e.target.id
		const index = e.target.getAttribute('index')
		const value = Number(e.target.value)
		const qty = isNaN(value) ? equipment[index].qty : value
		setEquipment(update(equipment, {
			[index]: {
				qty: { $set: qty },
				totalValue: { $set: calcItemTotalValue(props.items[id], qty) },
				totalWeight: { $set: calcItemTotalWeight(props.items[id], qty) }
			}
		}))
	}

	const handleEquip = (e) => {
		const id = e.target.id
		const category = e.target.name
		const alreadyEquipped = _.includes(equipped[category], id)
		if (category === 'weapons') {
			const weapons = alreadyEquipped ? _.without(equipped.weapons, id) : [...equipped.weapons, id]
			setEquipped(update(equipped, {
				weapons: { $set: weapons }
			}))
		} else {
			setEquipped(update(equipped, {
				[category]: { $set: alreadyEquipped ? null : id }
			}))
		}
	}

	return (
		<div className="container container--body">			
			<div className="section">
				<div className="grid--money">
					{denominations.map((denomination, i) => (
						<div className="grid--money__cell" key={`denomination${i}`}>
							<input
								type="number"
								id={denomination}
								data-testid={denomination}
								value={money[denomination]}
								onChange={(e) => setMoney({ ...money, [denomination]: Number(e.target.value) })}
							/>  
							<div>{denomination}</div>
						</div>
					))}
				</div>

				<div className="row--right">
					<h4 data-testid="totalMoney">Total money: {totalMoney} gp</h4> 
				</div>
			</div>
			
			<div className="grid--items">
				<h5 className="grid__col1">Item</h5>
				<h5>Qty</h5>
				<h5>Value</h5>
				<h5>Weight</h5>  

				{equipment.map((item, i) => (
					<Fragment key={i}>
						<button 
							className="grid__col1 button--link" 
							id={item.id}
							onClick={() => setSelected({ id: item.id, ...props.items[item.id] })}
						>
							{props.items[item.id].name}
						</button>                                
						<input 
							className="grid__col2" 
							type="number"
							id={item.id}
							data-testid={item.id}
							index={i}
							value={item.qty}
							onChange={(e) => handleChange(e)}
						/>                                
						<div className="grid__col3" data-testid={`${item.id}TotalValue`}>{item.totalValue} gp</div>
						<div className="grid__col4" data-testid={`${item.id}TotalWeight`}>{item.totalWeight} lbs</div> 
					</Fragment>
				))}

				<div className="grid__col1 grid--items__totals">Totals</div>
				<div className="grid__col3 grid--items__totals">{equipmentTotalValue} gp</div>
				<div className="grid__col4 grid--items__totals">{equipmentTotalWeight} lbs</div>

			</div>

			{/* <button>Add New Item</button> */}

			<ItemModal 
				selected={selected} 
				equipped={equipped}
				handleEquip={handleEquip}
				handleCloseModal={() => setSelected(undefined)}
			/>

		</div>
	)
}

const mapStateToProps = ({ profile, items }) => ({
	id: profile.id,
	money: profile.money,
	equipment: profile.equipment,
	equipped: profile.equipped,
	items
})
  
const mapDispatchToProps = (dispatch) => ({
	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})
  
  
export default connect(mapStateToProps, mapDispatchToProps)(EquipmentPage)