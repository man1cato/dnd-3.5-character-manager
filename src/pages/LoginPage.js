import React from 'react';
import {connect} from 'react-redux';
import {startLogin} from '../store/actions/auth';

export const LoginPage = ({startLogin}) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">D&D 3.5e Character Manager</h1>
            <button className="button--primary" onClick={startLogin}>Login with Google</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
