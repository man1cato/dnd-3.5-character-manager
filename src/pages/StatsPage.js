import React from 'react'
import { connect } from 'react-redux'
import { startEditProfile } from '../store/actions/profile'

import Abilities from '../components/Abilities'
import SkillSet from '../components/SkillSet'


export const StatsPage = ({ profile, jobClasses, startEditProfile }) => {
	const { xp, level } = profile
	const jobClass = jobClasses[profile.jobClass]
	const nextLevelXp = jobClass.levels[level + 1].xp

	const handleUpdate = (updates) => {
		startEditProfile(profile.id, updates)
	}

	return (
		<div className="container container--body">

			<div className="row">
				<h3>Level {profile.level}</h3>
				<span>XP: {xp}</span>
			</div>

			<div className="row">
				<span>Hit Die: {profile.hitDie}</span>
				<span>To next level: {nextLevelXp - xp}</span>
			</div>

			<div className="section">
				<Abilities
					abilities={profile.abilities}
					handleUpdate={handleUpdate}
				/>
			</div>
			
			<SkillSet 
				skillSet={profile.skillSet}
			/>

		</div>
	)
}

const mapStateToProps = (state) => ({
	profile: state.profile,
	jobClasses: state.jobClasses
})

const mapDispatchToProps = (dispatch) => ({ 
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})

export default connect(mapStateToProps, mapDispatchToProps)(StatsPage)
