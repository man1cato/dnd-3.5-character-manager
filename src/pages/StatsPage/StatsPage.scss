@import '../../styles/base/settings';


.StatsPage__row {
   display: flex;
   align-items: stretch;
   justify-content: space-between;
   margin-bottom: $m-size;
}