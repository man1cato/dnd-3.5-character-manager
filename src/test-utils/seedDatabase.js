import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}