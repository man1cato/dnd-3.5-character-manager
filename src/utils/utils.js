import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])