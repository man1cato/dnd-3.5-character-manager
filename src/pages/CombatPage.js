import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import PhysicalStats from '../components/PhysicalStats'
import AttackBonuses from '../components/AttackBonuses/AttackBonuses'
import Saves from '../components/Saves/Saves'
import Weapons from '../components/Weapons'
import PreparedSpells from '../components/PreparedSpells'
import { startEditProfile } from '../store/actions/profile'
import { calcAbilityMod } from '../utils/utils'


export const CombatPage = ({ profile, races, jobClasses, spells, startEditProfile }) => {
	const jobClassLevel = jobClasses[profile.jobClass].levels[profile.level]
	const race = races[profile.race]
	const handleUpdate = (updates) => {	startEditProfile(profile.id, updates) }

	return (
		<div className="container container--body">
			<div className="section">
				<h3 className="section__title">Physical Stats</h3>
				<PhysicalStats
					hp={profile.hp}
					ac={profile.ac}
					initBase={
						!!profile.abilities.dex.tempScore ? 
						calcAbilityMod(profile.abilities.dex.tempScore) : 
						calcAbilityMod(profile.abilities.dex.score)
					}
					initMod={profile.initMod}
					speed={race.speed}
					handleUpdate={handleUpdate}							
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Saving Throws</h3>
				<Saves
					saveMods={profile.saveMods}
					saveBases={jobClassLevel.saves}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">       
				<h3 className="section__title">Attack Bonuses</h3>
				<AttackBonuses 
					attackBonusMods={profile.attackBonusMods}
					abilities={profile.abilities}
					baseAttackBonuses={jobClassLevel.baseAttackBonuses}
					size={race.size}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Equipped Weapons</h3>
				<Weapons 
					weaponSet={profile.equipped.weapons}
					// meleeBonus={state.attacks.melee.total}
					// rangedBonus={state.attacks.ranged.total}
					// grappleBonus={state.attacks.grapple.total}
				/>
			</div>

			{profile.spellbook && (
				<div className="section">
					<h3 className="section__title">Prepared Spells</h3>
					<PreparedSpells 
						spellbook={profile.spellbook}
						spells={spells}
						handleUpdate={handleUpdate}
					/>
				</div>
			)}

		</div>
	)
}

const mapStateToProps = ({ profile, races, jobClasses, spells }) => ({
	profile,
	races,
	jobClasses,
	spells
})

const mapDispatchToProps = (dispatch) => ({
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)