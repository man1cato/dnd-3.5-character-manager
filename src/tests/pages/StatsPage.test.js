import React from 'react'
import { shallow } from 'enzyme'
import { render, fireEvent } from '@testing-library/react'

import { StatsPage } from '../../pages/StatsPage'
import { apiData } from '../utils/utils'
import { characterOne } from '../utils/seedDatabase'
 

const startEditProfile = jest.fn() 
let props

beforeAll(async () => {
	const api = await apiData()
	props = {
		profile: characterOne, 
		jobClasses: api.jobClasses,
		startEditProfile
	}
})


test('should render stats page with profile data', () => {
	const wrapper = shallow(<StatsPage {...props} />)
	expect(wrapper).toMatchSnapshot()
})
