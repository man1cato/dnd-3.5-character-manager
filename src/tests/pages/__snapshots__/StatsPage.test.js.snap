// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render stats page with profile data 1`] = `
<div
  className="container container--body"
>
  <div
    className="row"
  >
    <h3>
      Level 
      5
    </h3>
    <span>
      XP: 
      10000
    </span>
  </div>
  <div
    className="row"
  >
    <span>
      Hit Die: 
    </span>
    <span>
      To next level: 
      5000
    </span>
  </div>
  <div
    className="section"
  >
    <Abilities
      abilities={
        Object {
          "cha": Object {
            "score": 11,
          },
          "con": Object {
            "score": 16,
          },
          "dex": Object {
            "score": 18,
          },
          "int": Object {
            "score": 20,
          },
          "str": Object {
            "score": 14,
          },
          "wis": Object {
            "score": 14,
          },
        }
      }
      handleUpdate={[Function]}
    />
  </div>
  <Connect(SkillSet)
    skillSet={
      Array [
        Object {
          "id": "recfbplSsRmzNs68J",
          "ranks": 4,
        },
        Object {
          "id": "recoJ7dkAWtHAkj3v",
          "ranks": 3,
        },
        Object {
          "id": "recHCeKVn3CB0HcNA",
          "ranks": 5,
        },
        Object {
          "id": "recs2EXcqiCIfwO0w",
          "ranks": 9,
        },
      ]
    }
  />
</div>
`;
